src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs

```diff
@@ -297,5 +313,23 @@ private TCommandResult ProcessResponse(ConnectionId connectionId, CommandMessage
                 }
             }
         }
+
+        private bool ShouldAddTransientTransactionError(MongoException exception)
+        {
+            if (_session.IsInTransaction)
+            {
+                if (exception is MongoConnectionException)
+                {
+                    return true;
+                }
+
+                if (exception is MongoNotPrimaryException)
```

Comment:
Not sure why we don't return true for MongoNodeIsRecoveringException, but the tests don't require it.

Reply:
Do the tests require this?  Java driver only adds this for socket exceptions.

> Any command error that includes the “TransientTransactionError” error label in the “errorLabels” field. Any network error encountered running any command other than commitTransaction in a transaction. If a network error occurs while running the commitTransaction command then it is not known whether the transaction committed or not, and thus the “TransientTransactionError” label MUST NOT be added.

Reply:
Returning true for MongoNotPrimaryException is required by:

    error-labels.json:NotMaster errors contain transient label

Reply:
I think you don't need this in Java because the Java NotMasterException is a subclass of MongoCommandException and the reply from the server already contains the required TransientTransactionError error label.

To be able to remove this line from the C# driver I would have to go through with the plan to change the base class of MongoNotMasterException in MongoNodeIsRecoveringException to MongoCommandException.

Reply:
Ack.  I think that should be done eventually, as we could end up missing additional error labels returned from the server

Reply:
Look like you're doing this in another PR.

Reply:
Right.

I will rebase this work on that PR.

I wanted to separate the refactoring of the exceptions from the transaction work.